### ISSUE TITLE
Better project documentation

### ISSUE THREAD

**Issue by wiztigers** [TypeCobolTeam/TypeCobol#166]
Our current project welcome page seems a bit daunting : we're not sure throwing license and not-so-up-to-date technical details at the face of a new user is the better thing to do.

This asks too the question about the front page `README.md` size, which will become too large the more information we add to it.

Lastly, information about the project is currently scattered accross
* the GitHub welcome page
* our [project wiki](https://github.com/TypeCobolTeam/TypeCobol/wiki) (kind poorly organized for now)
* [this directory](https://github.com/TypeCobolTeam/TypeCobol/tree/master/TypeCobol/Documentation) (containing bloated .xls or .docx documents not versionning-friendly at all)

What is the best way to improve this, to lower our project entry bareer ?

**Comment by wiztigers** [comment]
### 1st lead : write a user manual

Let's begin with this for the sake of completeness : writing a .pdf or .doc user manual.

What's wrong with this ?
* this kind of document are *not* VCS friendly : such binary formats don't allow proper tracking of who did the modification, as long and when and why
* if we use a proprietary format, the lasting quality of the documentation is not guaranteed. Talk me about ms office retrocompatibility.
* seriously, we're 2016 now. We've got lots of tools to take textual input in one hand, formatting (stylesheets, ...) in the other, and combine all this to get different documentations in different formats (.pdf, .html and so on) suiting different needs. No need to merge content and formatting in the same file anymore.

We may however want such documentation eventually.
But it would be generated instead of being the only thing we would have.

**Comment by wiztigers** [comment]
### 2nd lead : documentation markdown files

The GitHub welcome page is two things : a source hierarchy view as long as the contents of the root `README.md` file, displayed as a web page.
Moreover, every `.md` markdown file can be displayed as a web page if viewed via a browser navigating through the sources on the github website.

We could thus write all our documentation in markdown files. They would site next to the sources and be viewed in a user friendly way in the browser, very similarly navigating through the wiki.

**PROS**
* no versionning problem : this documentation would sit in the sources, being versionned like all other committed files. Relative links betweed markdown files [just work](https://github.com/blog/1395-relative-links-in-markup-files) ; in other word, navigating through a link containted in a old version of a file to another markdown file would display this file *as it was at this old version of the project*. Depending on how the link would be written, it could of course always link to the HEAD version of a file.
* no lisibility problem : nicely displayed through github as a webpage, markdown files are just perfectly user readable text files.

**CONS**
* Even if GitHub allows someone to edit a markdown file directly through the browser, outside of it, text files can be "grim" to work with to some users. A wiki would be better in this case.

**Comment by wiztigers** [comment]
### 3rd lead : full GitHub wiki

We could just rethink our current wiki organization, put all the technical information currently in the welcome page in the wiki, point to the wiki in the welcome page and call it a day.
GitHub wiki uses markdown syntax, allowing a clear standardized way of displaying info.

There are no problem concerning the lasting quality of a github wiki, as each one is just another github wiki:
```
Each wiki is a Git repository, so you're able to push and pull them like anything else.
Each wiki respects the same permissions as the source repository.
Just add ".wiki" to any repository name in the URL, and you're ready to go.
```
In other words, one can just `git clone https://github.com/TypeCobolTeam/TypeCobol.wiki.git TypeCobolWiki/` and they have the wiki ; then tasks and scripts can be run on it, for example to produce end-user standalone documentation like .pdf file or whatever suits you.

**PROS**
* All the advantages of the 2nd lead just up there
* User-friendly wiki editing, from about anywhere
* Separated from the sources. May avoid source clutter.

**CONS**
* Separated from the sources. Prevents source+documentation update in a single shot.

**Comment by wiztigers** [comment]
Ooops, I forgot this one : 

### 4th lead : project website

As you can imagine, we can create a sexy website, including all kinds of modern html5 and css and, you know ... [juicy stuff](https://www.youtube.com/watch?v=Fy0aCDmgnxg). (≧ω≦*)

To be up and running fast, github provides us a way to host such a website called GitHub pages ([presentation](https://pages.github.com/) and [more details](https://help.github.com/categories/github-pages-basics/)). This solution includes website generation tools.

**PROS**
* custom content !
* eye candy !

**CONS**
* yet another thing to develop !

**Comment by wiztigers** [comment]
@prudholu @vavans @osmedile 
Clean project presentation is a pivotal tool of community building, so thanks in advance for your feedback !

**Comment by laurentprudhon** [comment]
First proposition of a plan for the future documentation : 
Introduction to the TypeCobol initiative (why, what, unique features, how / license, roadmap overview)

* Part 1 – Concepts and syntax for the users of the TypeCobol language & editor

TypeCobol developer guide
                Common scenarios and code samples
                TypeCobol Editor manual

Cobol concepts (user friendly presentation of information that is today deeply buried and obfuscated in the IBM Cobol reference)
                Source text
                               Programs, memory, calls
                               Files, libraries, copy
                               Compilation directives vs statements
                               Text format, line types, continuations
                               Replace, copy replacing
                               Text configuration statements
                Program structure
                               Identification
                               Configuration
                                               Main parameters
                                               File access
                               Data description
                                               Data structures
                               Procedures
                                               Control flow
                                               Code statements
                Data model
                               Data section, memory model
                               Picture string, type system
                               Move and display
                Statements - by theme
                               Doc and precise grammar for each statement
                Expressions
                               Expressions grammar

TypeCobol concepts
                Types
                Interfaces
                …

TypeCobol language syntax reference
                ... tables for each feature, linked to source code ...

* Part 2 – Concepts and syntax for the users of the TypeCobol library and analysis tools

TypeCobol library user guide
                Common scenarios and code samples

TypeCobol compiler architecture
                Main phases
                File
                               Find on disk
                               Encoding
                               End of lines
                Text
                                               
                Scanner
                Preprocessor
                Code elements parser
                Program class parser
                Type checker
                Error handling

TypeCobol library – end user API reference
                ... tables for each class useful for simple code analysis ..

* Part 3 – Concepts and syntax for the extension writers

TypeCobol extension writer guide
                Common scenarios and code samples

TypeCobol extension model and concepts

TypeCobol API reference – extension writer version
                ... tables for each class ...

* part 4 - 

TypeCobol contributor guide
                Common scenarios : documentation and procedures

Architecture - Internal details

Algorithms for Incremental and Multithread compilation
                Incremental challenges
                               - for File
                               - for Text
                Multithread challenges
                               - for File
                               - for Text
                Solutions applied
                               - for file
                               - general document

TypeCobol API reference - contributor version
                ... tables for each class ...

**Comment by wiztigers** [comment]
Okay, I've [committed](https://github.com/TypeCobolTeam/TypeCobol/wiki/_Sidebar/bce6a0c5f8794edfe06f62ba68d35d7dd3d7372d) a new sidebar, merging your idea of a table of content with the one I had in mind.

It may be a bit daunting at the moment. A table content height shouldn't be so high ; at best it shouldn't exceed the page height IMHO. I created like this however so anyone can create a page simply by clicking on the corresponding "red" link in the sidebar -and this sidebar will follow the pages naming scheme I have in mind.

This sidebar appears on all wiki pages, for now.

Feel free to fill up the blanks !
That's what I'll do in the next weeks.

And tell me if there's a problem with the current plan.

**Comment by wiztigers** [comment]
This issue is so vague.
Table of content has been implemented months ago.
Although a lot remains to be done (fill in blanks, update readme, update parsing steps, ...), each should be its own issue.

Plus, this is a meta-issue that is better represented by the **Wiki improvement** milestone.
